### PR TITLE
[MAILET] Allow chaining composite matcher

### PR DESCRIPTION
**Pull request by quantranhong1999**
```
<matcher name="relay-allowed" match="org.apache.james.mailetcontainer.impl.matchers.Or">
    <matcher match="SMTPAuthSuccessful"/>
    <matcher match="SMTPIsAuthNetwork"/>
    <matcher match="SentByMailet"/>
    <matcher match="org.apache.james.jmap.mailet.SentByJmap"/>
</matcher>
<matcher name="should-check-spam" match="org.apache.james.mailetcontainer.impl.matchers.Not">
    <matcher match="relay-allowed"/>
</matcher>
```
was not possible before.

Loading the matcher from composite matchers collection also, not only using class loader solves the issue.

### CODE REVIEW

File: server/mailet/mailetcontainer-impl/src/test/java/org/apache/james/mailetcontainer/impl/matchers/InvertMatcherWithCompositeMatcherTest.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        //  </matcher>
+
+        when(matcher1.match(mail)).thenReturn(List.of());
+        when(matcher2.match(mail)).thenReturn(List.of());
+        orTestee.add(matcher1);
+        orTestee.add(matcher2);
+        MatcherInverter inverterMatcher = new MatcherInverter(orTestee);
+
+        assertThat(inverterMatcher.match(mail)).containsExactlyInAnyOrder(recipient1, recipient2);
+    }
```
Comment (quantranhong1999) [inline server/mailet/mailetcontainer-impl/src/test/java/org/apache/james/mailetcontainer/impl/matchers/InvertMatcherWithCompositeMatcherTest.java:156]: cc @chibenwa likely this ~ NOR is what you wanted?

```java
<matcher name="relay-allowed" match="org.apache.james.mailetcontainer.impl.matchers.Or">
    <matcher match="SMTPAuthSuccessful"/>
    <matcher match="SMTPIsAuthNetwork"/>
    <matcher match="SentByMailet"/>
    <matcher match="org.apache.james.jmap.mailet.SentByJmap"/>
</matcher>
<matcher name="should-check-spam" match="org.apache.james.mailetcontainer.impl.matchers.Not">
    <matcher match="relay-allowed"/>
</matcher>
```
<=>
```java
<matcher name="should-check-spam" notmatch="org.apache.james.mailetcontainer.impl.matchers.Or">
    <matcher match="SMTPAuthSuccessful"/>
    <matcher match="SMTPIsAuthNetwork"/>
    <matcher match="SentByMailet"/>
    <matcher match="org.apache.james.jmap.mailet.SentByJmap"/>
</matcher>
```
?

File: server/mailet/mailetcontainer-impl/src/main/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessor.java [hunk 2]
```diff
@@ -343,6 +336,18 @@ private void parseConfiguration() throws MessagingException, ConfigurationExcept
...
+    private Matcher loadMatcher(Map<String, Matcher> compositeMatchers, String matcherName) throws MessagingException {
+        Matcher matcher;
+        // try to load from compositeMatchers first
+        matcher = compositeMatchers.get(matcherName);
+        if (matcher == null) {
+            // no composite Matcher found, try to load it via
+            // MatcherLoader
+            matcher = matcherLoader.getMatcher(createMatcherConfig(matcherName));
+        }
+        return matcher;
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-impl/src/main/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessor.java:348]: Use optional and avoid variable allocation?

**Review by chibenwa — COMMENTED** [review]
Good work.

=> Don't forget to create a JIRA ticket?

I like this `notMatch` tag.

How about being able to write something like:

```
            <mailet notMatch="RecipientIs=error@domain.tld" class="org.apache.james.Whatever"/>
```

directly too?

**Review by Arsnael — COMMENTED** [review]
Read it nothing to add

**Comment by quantranhong1999** [comment]
I rebased & reworded commits to include the JIRA ticket. Two latest commits resolve review comments.

File: server/apps/memory-app/src/test/resources/mailetcontainer.xml [hunk 3]
```diff
@@ -52,6 +52,22 @@
...
+                <matcher match="SentByMailet"/>
+                <matcher match="org.apache.james.jmap.mailet.SentByJmap"/>
+            </matcher>
+            <matcher name="chaining-matcher" match="org.apache.james.mailetcontainer.impl.matchers.Not">
+                <matcher match="relay-allowed"/>
+            </matcher>
+            <matcher name="chaining-not-matcher" match="org.apache.james.mailetcontainer.impl.matchers.Not">
+                <matcher notmatch="relay-allowed"/>
+            </matcher>
+            <mailet notmatch="chaining-not-matcher" class="org.apache.james.transport.mailets.debug.Counter"/>
```
Comment (chibenwa) [inline server/apps/memory-app/src/test/resources/mailetcontainer.xml:69]: Let's document `notmatch` into `docs/modules/servers/partials/configure/mailetcontainer.adoc`

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**